program/src/error.rs

```diff
@@ -0,0 +1,42 @@
+use num_derive::FromPrimitive;
+use solana_program::{
+    decode_error::DecodeError,
+    msg,
+    program_error::{PrintProgramError, ProgramError},
+};
+use thiserror::Error;
+
+#[derive(Error, Clone, Debug, Eq, PartialEq, FromPrimitive)]
+pub enum StakeError {}
+
+impl PrintProgramError for StakeError {
+    fn print<E>(&self) {
+        msg!(&self.to_string());
+    }
+}
+
+impl From<StakeError> for ProgramError {
+    fn from(e: StakeError) -> Self {
+        ProgramError::Custom(e as u32)
+    }
+}
+
+impl<T> DecodeError<T> for StakeError {
+    fn type_of() -> &'static str {
+        "Stake Error"
+    }
+}
+
+#[macro_export]
+macro_rules! err {
+    ( $error:expr ) => {{
+        Err($error.into())
+    }};
+    ( $error:expr, $msg:expr ) => {{
+        solana_program::msg!("[ERROR] {}", $msg);
+        Err($error.into())
+    }};
+    ( $error:expr, $msg:literal, $($args:tt)+ ) => {{
+        err!($error, &format!($msg, $($args)+))
+    }};
+}
```

Comment:
This is interesting. It's not included in the SPL Program Error macro, but your entrypoint setup combined with `PrintProgramError` should already do most of this, right?